src/NServiceBus.Core/Transports/TransportReceiver.cs

```diff
@@ -61,19 +64,42 @@ class TransportReceiver
 
         async Task InvokePipeline(PushContext pushContext)
         {
+            //note: we should get this value from the transport in the future to make it more accurate
+            var pipelineStartedAt = DateTime.UtcNow;
+
+
             using (var childBuilder = builder.CreateChildBuilder())
             {
                 var rootContext = new RootContext(childBuilder, pipelineCache, eventAggregator);
-                var context = new TransportReceiveContext(new IncomingMessage(pushContext.MessageId, pushContext.Headers, pushContext.BodyStream), pushContext.TransportTransaction, pushContext.ReceiveCancellationTokenSource, rootContext);
+
+                //notifications are only relevant for the main pipeline since satellites pipelines are not exposed to user in any way
+                if (isMainReceiver)
+                {
+                    var sw = new Stopwatch();
+
+                    sw.Start();
+
+                    // float the sw via the context to avoid threading issues should the continuation where we access 
+                    // it is executed on a different thread
+                    rootContext.Set(sw);
```

Comment:
I don't understand what threading issues can be caused by using a local variable and how floating it through context solves this. Do you have some more background info for me about this?